lib/CodeGen/CGExprScalar.cpp

```diff
@@ -576,6 +576,18 @@ class ScalarExprEmitter
     return EmitNullValue(E->getType());
   }
   Value *VisitExplicitCastExpr(ExplicitCastExpr *E) {
+    // TODO: HCC specific kludge due to how ROCDL feels it is OK to force
+    //       address spaces unto everyone.
+    if (auto CCE = dyn_cast_or_null<CStyleCastExpr>(E)) {
+      auto DstTy = ConvertType(CCE->getTypeAsWritten());
+      auto SrcTy = ConvertType(CCE->getSubExpr()->getType());
+
+      if (DstTy->isPointerTy() && SrcTy->isPointerTy() &&
+          DstTy->getPointerAddressSpace() != SrcTy->getPointerAddressSpace())
+        if (CCE->getCastKind() != CK_AddressSpaceConversion)
+          CCE->setCastKind(CK_AddressSpaceConversion);
+    }
+
```

Comment:
I'm just wondering why we need this now since we've been using rocdl forever and is this hcc specific?

Reply:
You have, but:

- you never do Debug builds so you don't see asserts firing;
- C-casts used to do an address-space cast but they no longer do in C++ AFAICT, which makes most of the noise in the math lib which casts to private (5) be silently incorrect, as you get bitcasts.